### PR TITLE
Heroku app review

### PR DESCRIPTION
**Pull request by alan**
Makes use of the `app.json` hooks to set up a new pvb2 app in SSO when it gets deployed for the first time and cleaning up after the PR gets merged or closed.

### CODE REVIEW

File: lib/tasks/heroku.rake [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    sso_url = ENV.fetch('MOJSSO_URL')
+
+    new_app_name = ENV.fetch('HEROKU_APP_NAME')
+
+    service_uri = "https://#{new_app_name}.herokuapp.com"
+
+    post_data = {
+      'app_name' => 'Prison Visits Booking',
+      'new_app_name' => new_app_name,
```
Comment (mloughran) [inline lib/tasks/heroku.rake]: What about including 'Prison Visit Booking' in the new app name? Just thinking the raw heroku app name might be confusing for users. Could be something like

    Prison Visit Booking (review app: pvb-staff-pr-454)

File: lib/tasks/heroku.rake [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    require 'excon'
+
+    sso_url = ENV.fetch('MOJSSO_URL')
+
+    new_app_name = ENV.fetch('HEROKU_APP_NAME')
+
+    service_uri = "https://#{new_app_name}.herokuapp.com"
+
+    post_data = {
+      'app_name' => 'Prison Visits Booking',
```
Comment (mloughran) [inline lib/tasks/heroku.rake]: Should we make the app name configurable? It will vary by environment.

Comment (alan) [inline lib/tasks/heroku.rake]: We can, I assumed Staging and Production would have the same app name for PVB. How come they are different?

File: README.md [hunk 3]
```diff
@@ -225,6 +225,11 @@ generate a secure secret key.
 This is used to build public links in emails. It must be set in the production
 environment to `https://prisonvisits.service.gov.uk/`.
 
+#### `SSO_BASIC_USER`, `SSO_BASIC_PASSWORD`, `SSO_REVIEW_APP_ID`
```
Comment (mloughran) [inline README.md]: Maybe `SSO_REVIEW_BASIC_USER`, `SSO_REVIEW_PARENT_ID`

File: lib/tasks/heroku.rake [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  desc 'Runs once on the first PR deploy to Heroku, used in app.json'
+  task post_deploy: :environment do
+    app_name = ENV.fetch('HEROKU_APP_NAME')
+
+    original_app_id = ENV.fetch('SSO_REVIEW_APP_ID')
+
+    service_uri = "https://#{app_name}.herokuapp.com"
+
+    post_data = {
+      'app_id' => original_app_id,
```
Comment (mloughran) [inline lib/tasks/heroku.rake]: Decided to rename to `parent_app_id`

File: lib/tasks/heroku.rake [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      'new_app_name' => Rails.configuration.sso_app_id,
```
Comment (mloughran) [inline lib/tasks/heroku.rake]: Decided to rename to `new_app_id`

File: lib/tasks/heroku.rake [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      expects: 200,
+      idempotent: true,
+      body: post_data.to_json,
+      headers: { 'Content-Type' => 'application/json' }
+    )
+  end
+
+  desc 'Runs once after the PR is merged or closed, used in app.json'
+  task pr_destroy: :environment do
+    fail('not a review app') unless ENV['HEROKU_APP_NAME']
```
Comment (mloughran) [inline lib/tasks/heroku.rake]: If possible we should check for the parent app name env variable.

File: config/initializers/omniauth.rb [hunk 7]
```diff
@@ -1,13 +1,31 @@
...
+# Used for Heroku review apps.
+# SSO app must have a unique id, so when a new heroku review app gets deployed
+# we use the name of the heroku app to be the id of a duplicate pvb app.
+heroku_app_name = ENV.fetch('HEROKU_APP_NAME', nil)
+review_app_id = if heroku_app_name
+                  # Same value as used in the heroku rake task
+                  "Prison Visits Booking (review app: #{heroku_app_name})"
+                end
+
+# Heroku review apps don't inherit the MOSSO_ID environment variable.
```
Comment (StupidCodeFactory) [inline config/initializers/omniauth.rb]: little typo

**Pull request merged**